### PR TITLE
Tif/cross check reporting

### PR DESCRIPTION
**Pull request by jcurtisk**
## What does this PR do?

This PR adds support for discovering and reporting the unlikely edge case where TIAF Python test selection does not select a test that would have failed. This reporting will enable stakeholders to locate errors in TIAF selection logic as well as identify flakey tests.

## How was this PR tested?

This PR was manually tested.

### CODE REVIEW

File: Code/Tools/TestImpactFramework/Runtime/Native/Code/Source/TestRunner/Native/TestImpactNativeErrorCodeChecker.cpp [hunk 1]
```diff
@@ -6,27 +6,10 @@
  *
  */
 
-#include <TestEngine/Native/TestImpactNativeErrorCodeChecker.h>
+#include <TestRunner/Native/TestImpactNativeErrorCodeChecker.h>
 
 namespace TestImpact
 {
-    // Known error codes for test runner and test library
```
Comment (jcurtisk) [inline Code/Tools/TestImpactFramework/Runtime/Native/Code/Source/TestRunner/Native/TestImpactNativeErrorCodeChecker.cpp:13]: Code moved

File: Code/Tools/TestImpactFramework/Runtime/Native/Code/Source/TestRunner/Native/TestImpactNativeErrorCodeChecker.h [hunk 2]
```diff
@@ -15,6 +15,23 @@
 
 namespace TestImpact
 {
+    // Known error codes for test runner and test library
```
Comment (jcurtisk) [inline Code/Tools/TestImpactFramework/Runtime/Native/Code/Source/TestRunner/Native/TestImpactNativeErrorCodeChecker.h:18]: Code move

File: Code/Tools/TestImpactFramework/Runtime/Python/Code/Source/TestRunner/Python/TestImpactPythonErrorCodeChecker.cpp [hunk 3]
```diff
@@ -6,29 +6,10 @@
  *
  */
 
-#include <TestEngine/Python/TestImpactPythonErrorCodeChecker.h>
+#include <TestRunner/Python/TestImpactPythonErrorCodeChecker.h>
 
 namespace TestImpact
 {
-    // Known error codes for Python and PyTest library
```
Comment (jcurtisk) [inline Code/Tools/TestImpactFramework/Runtime/Python/Code/Source/TestRunner/Python/TestImpactPythonErrorCodeChecker.cpp:13]: Code move

File: Code/Tools/TestImpactFramework/Runtime/Python/Code/Source/TestRunner/Python/TestImpactPythonErrorCodeChecker.h [hunk 4]
```diff
@@ -15,6 +15,25 @@
 
 namespace TestImpact
 {
+    // Known error codes for Python and PyTest library
```
Comment (jcurtisk) [inline Code/Tools/TestImpactFramework/Runtime/Python/Code/Source/TestRunner/Python/TestImpactPythonErrorCodeChecker.h:18]: Code move

File: Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Include/Static/TestImpactCommandLineOptionsUtils.h [hunk 5]
```diff
@@ -30,26 +30,21 @@ namespace TestImpact
...
     template<typename T>
     AZStd::optional<T> ParseBinaryStateOption(
-        const AZStd::string& optionName,
-        const AZStd::pair<OptionValue<T>,
-        OptionValue<T>>& state, const AZ::CommandLine& cmd)
+        const AZStd::string& optionName, const AZStd::pair<OptionValue<T>, OptionValue<T>>& state, const AZ::CommandLine& cmd)
     {
-        if (const auto numSwitchValues = cmd.GetNumSwitchValues(optionName);
-            numSwitchValues)
+        if (const auto numSwitchValues = cmd.GetNumSwitchValues(optionName); numSwitchValues)
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Include/Static/TestImpactCommandLineOptionsUtils.h:35]: suggestion: Rather than relying on an implicit `0` check here, having this be `numSwitchValues == 0` would be clearer (repeated in a few other spots)

File: Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Include/Static/TestImpactCommandLineOptionsUtils.h [hunk 6]
```diff
@@ -97,14 +88,22 @@ namespace TestImpact
...
     template<typename T>
-    AZStd::optional<T> ParseAbortContinueOption(const AZStd::string& optionName, const AZStd::pair<T, T>& states, const AZ::CommandLine& cmd)
+    AZStd::optional<T> ParseAbortContinueOption(
+        const AZStd::string& optionName, const AZStd::pair<T, T>& states, const AZ::CommandLine& cmd)
     {
-        return ParseBinaryStateOption(optionName, BinaryStateOption<T>{ {"abort", states.first}, { "continue", states.second } }, cmd);
+        return ParseBinaryStateOption(optionName, BinaryStateOption<T>{ { "abort", states.first }, { "continue", states.second } }, cmd);
+    }
+
+    //! Attempts to pass a specialization of the binary state option where the command lien values are "live" and "null".
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Include/Static/TestImpactCommandLineOptionsUtils.h:102]: update: lien -> line

File: Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp [hunk 7]
```diff
@@ -495,6 +510,8 @@ namespace TestImpact
             "                                                                production targets in the dependency graph(if no dependency graph data \n"
             "                                                                available, no prioritization will occur).\n"
             "    -safemode=<on,off>                                          Flag to specify a safe mode sequence where the set of unselected \n"
+            "    -testrunner=<live,null>                                     Whether to use the null test runner (on) or run the tests(off). \n"
+            "                                                                If not set, defaults to running tests.                          \n"
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp:514]: nit: running 'the' tests

File: Code/Tools/TestImpactFramework/Frontend/Console/Python/Code/Source/TestImpactPythonCommandLineOptions.cpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#include <TestImpactFramework/TestImpactUtils.h>
+
+#include <AzCore/Settings/CommandLine.h>
+
+namespace TestImpact
+{
+    namespace
+    {
+    } // namespace
+} // namespace TestImpact
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Python/Code/Source/TestImpactPythonCommandLineOptions.cpp:20]: question: Can this file just be removed? Nothing going on in here it seems

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h [hunk 9]
```diff
@@ -37,6 +37,7 @@ namespace TestImpact
             //! @param duration The duration that this test run took to complete.
             //! @param result The result of the run.
             TestRunBase(
+                const AZStd::string& test_namespace,
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h:40]: coding standards: `test_namespace` -> `testNamespace`

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h [hunk 10]
```diff
@@ -48,6 +49,9 @@ namespace TestImpact
             //! Returns the test target name.
             const AZStd::string& GetTargetName() const;
 
+            //! Returns the test target name space.
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h:52]: nit: `name space` -> `namespace`

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h [hunk 11]
```diff
@@ -142,7 +147,8 @@ namespace TestImpact
                 AZStd::chrono::high_resolution_clock::time_point startTime,
                 AZStd::chrono::milliseconds duration,
                 TestRunResult result,
-                AZStd::vector<Test>&& tests);
+                AZStd::vector<Test>&& tests,
+                const AZStd::string& test_namespace);
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Headers/TestImpactFramework/TestImpactClientTestRun.h:151]: standards: testNamespace

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/TestImpactClientTestRun.cpp [hunk 12]
```diff
@@ -120,8 +127,9 @@ namespace TestImpact
             AZStd::chrono::high_resolution_clock::time_point startTime,
             AZStd::chrono::milliseconds duration,
             TestRunResult result,
-            AZStd::vector<Test>&& tests)
-            : TestRunBase(name, commandString, startTime, duration, result)
+            AZStd::vector<Test>&& tests,
+            const AZStd::string& test_namespace)
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/TestImpactClientTestRun.cpp:131]: standards: test_namespace -> testNamespace

**Review by hultonha — APPROVED** [review]
Overall change looks good to me, just a few small things to fix but I'm happy to approve for now

File: Code/Tools/TestImpactFramework/Frontend/Console/Python/Code/Source/TestImpactPythonCommandLineOptions.cpp (hunk 8 above)
Comment (jcurtisk) [reply Code/Tools/TestImpactFramework/Frontend/Console/Python/Code/Source/TestImpactPythonCommandLineOptions.cpp:20]: Will remove now! There was some logic here but it got refactored out and I forgot to remove this.

Comment (hultonha) [reply Code/Tools/TestImpactFramework/Frontend/Console/Python/Code/Source/TestImpactPythonCommandLineOptions.cpp]: Thanks!

**Review by jonawals — APPROVED** [review]
